File: ts-route-plan-service/internal/security/jwt.go
package security

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"strings"
	"time"
)

// ValidateBearer reproduces the local HS256 validation used by ts-common JWTFilter.
// A non-Bearer Authorization value is treated exactly like no token.
func ValidateBearer(authorization string, now time.Time) error {
	if !strings.HasPrefix(authorization, "Bearer ") {
		return nil
	}
	token := strings.TrimPrefix(authorization, "Bearer ")
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return errors.New("malformed JWT")
	}
	headerBytes, err := base64.RawURLEncoding.DecodeString(parts[0])
	if err != nil {
		return err
	}
	var header struct {
		Algorithm string `json:"alg"`
	}
	if err := json.Unmarshal(headerBytes, &header); err != nil || header.Algorithm != "HS256" {
		return errors.New("unsupported JWT")
	}
	signature, err := base64.RawURLEncoding.DecodeString(parts[2])
	if err != nil {
		return err
	}
	mac := hmac.New(sha256.New, []byte("secret"))
	mac.Write([]byte(parts[0] + "." + parts[1]))
	if !hmac.Equal(signature, mac.Sum(nil)) {
		return errors.New("invalid JWT signature")
	}
	payloadBytes, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return err
	}
	var payload struct {
		Exp json.Number `json:"exp"`
	}
	decoder := json.NewDecoder(strings.NewReader(string(payloadBytes)))
	decoder.UseNumber()
	if err := decoder.Decode(&payload); err != nil {
		return err
	}
	if payload.Exp == "" {
		return errors.New("JWT has no expiration")
	}
	exp, err := payload.Exp.Int64()
	if err != nil {
		return err
	}
	if time.Unix(exp, 0).Before(now) {
		return errors.New("JWT expired")
	}
	return nil
}

File: ts-travel-plan-service/app/security/jwt.py
from __future__ import annotations

import base64
import hashlib
import hmac
import json
import time


def validate_optional_bearer(authorization: str | None, now: float | None = None) -> bool:
    if not authorization or not authorization.startswith("Bearer "):
        return True
    token = authorization[7:]
    try:
        header_part, payload_part, signature_part = token.split(".")
        header = json.loads(_decode(header_part))
        payload = json.loads(_decode(payload_part))
        if header.get("alg") != "HS256":
            return False
        expected = hmac.new(b"secret", f"{header_part}.{payload_part}".encode(), hashlib.sha256).digest()
        signature = base64.urlsafe_b64decode(_pad(signature_part))
        expiration = payload["exp"]
        return hmac.compare_digest(signature, expected) and float(expiration) * 1000 >= (now or time.time()) * 1000
    except (KeyError, TypeError, ValueError, json.JSONDecodeError):
        return False


def _decode(value: str) -> str:
    return base64.urlsafe_b64decode(_pad(value)).decode()


def _pad(value: str) -> str:
    return value + "=" * (-len(value) % 4)
